internal/util/logging/mongo_handler.go

```diff
@@ -0,0 +1,227 @@
+// Copyright 2021 FerretDB Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package logging
+
+import (
+	"bytes"
+	"context"
+	"io"
+	"log/slog"
+	"maps"
+	"runtime"
+	"slices"
+	"strconv"
+	"sync"
+
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/bson/primitive"
+
+	"github.com/FerretDB/FerretDB/v2/internal/util/must"
+)
+
+// mongoHandler is a [slog.Handler] that writes logs by using mongo structured JSON format.
+// The format returns log entries with Relaxed Extended JSON specification.
+// The format is not stable.
+//
+//nolint:vet // for readability
+type mongoHandler struct {
+	opts *NewHandlerOpts
+
+	ga        []groupOrAttrs
+	testAttrs map[string]any
+
+	m   *sync.Mutex
+	out io.Writer
+}
+
+// MongoLog represents a single log message in mongo structured JSON format.
+// For now some fields are ignored and may be empty.
+//
+//nolint:vet // to preserve field ordering
+type MongoLog struct {
+	Timestamp primitive.DateTime `bson:"t"`
```

Comment:
Do we need to expose `primitive.DateTime` to the caller? Isn't the encoder support `time.Time`?

Also, "t" likely means time, not timestamp (which is a separate BSON/ExtJSON type).

Reply:
Both `primitive.DateTime`, and "Timestamp" are used, based on the structured logging documentation:

```
{
"t": <Datetime>, // timestamp
"s": <String>, // severity
"c": <String>, // component
"id": <Integer>, // unique identifier
"ctx": <String>, // context
"svc": <String>, // service
"msg": <String>, // message body
"attr": <Object>, // additional attributes (optional)
"tags": <Array of strings>, // tags (optional)
"truncated": <Object>, // truncation info (if truncated)
"size": <Object> // original size of entry (if truncated)
}
```

As `time.Time` marshals to `primitive.DateTime`, I've changed the type in the struct.